### ISSUE TITLE
Adding hilt to existing Android App -- "This feature requires ASM7" and "com.android.tools.r8.errors.a: zip file closed"

### ISSUE THREAD

**Issue by Allen-HiCharlie** [google/dagger#2024]
Hey Dagger team, I'm adding hilt to my existing Android app (not a new project) and I'm seeing the following errors:

```
Caused by: java.lang.UnsupportedOperationException: This feature requires ASM7
```

And

```
zip file closed
Stack trace:
com.android.tools.r8.errors.a: zip file closed
	at com.android.tools.r8.ir.conversion.O.b(:57)
	at com.android.tools.r8.ir.conversion.O.a(:104)
	at com.android.tools.r8.ir.conversion.O.a(:53)
	at com.android.tools.r8.graph.C.b(:43)
	at com.android.tools.r8.ir.conversion.O.b(:35)
	at com.android.tools.r8.utils.U0.a(:10)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:322)
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134)
	at com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:66)
	at com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:36)
	at com.android.tools.r8.utils.U0.a(:3)
	at com.android.tools.r8.ir.conversion.O.a(:10)
	at com.android.tools.r8.D8.d(:29)
	at com.android.tools.r8.D8.b(:1)
	at com.android.tools.r8.utils.W.a(:30)
	at com.android.tools.r8.D8.run(:11)
	at com.android.builder.dexing.D8DexArchiveBuilder.convert(D8DexArchiveBuilder.java:116)
	at com.android.build.gradle.internal.dexing.DexWorkActionKt.process(DexWorkAction.kt:194)
	at com.android.build.gradle.internal.dexing.DexWorkActionKt.processNonIncrementally(DexWorkAction.kt:164)
	at com.android.build.gradle.internal.dexing.DexWorkActionKt.launchProcessing(DexWorkAction.kt:91)
	at com.android.build.gradle.internal.dexing.DexWorkAction.run(DexWorkAction.kt:49)
	at org.gradle.workers.internal.AdapterWorkAction.execute(AdapterWorkAction.java:57)
	at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.create(NoIsolationWorkerFactory.java:66)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.create(NoIsolationWorkerFactory.java:62)
	at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:98)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1.lambda$execute$0(NoIsolationWorkerFactory.java:62)
	at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
	at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.workers.internal.AbstractWorker.executeWrappedInBuildOperation(AbstractWorker.java:41)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1.execute(NoIsolationWorkerFactory.java:59)
	at org.gradle.workers.internal.DefaultWorkerExecutor.lambda$submitWork$2(DefaultWorkerExecutor.java:198)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:215)
	at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:164)
	at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:131)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: java.util.concurrent.ExecutionException: com.android.tools.r8.errors.a: zip file closed
		at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:552)
		at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:513)
		at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:86)
		at com.android.tools.r8.utils.U0.a(:14)
		at com.android.tools.r8.utils.U0.a(:9)
		... 40 more
	[CIRCULAR REFERENCE:com.android.tools.r8.errors.a: zip file closed]
Caused by: java.lang.IllegalStateException: zip file closed
	at java.util.zip.ZipFile.ensureOpen(ZipFile.java:686)
	at java.util.zip.ZipFile.getEntry(ZipFile.java:315)
	at com.android.tools.r8.ArchiveClassFileProvider.getProgramResource(:2)
	at com.android.builder.dexing.r8.CachingArchiveClassFileProvider.lambda$getProgramResource$0(CachingArchiveClassFileProvider.java:38)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1688)
	at com.android.builder.dexing.r8.CachingArchiveClassFileProvider.getProgramResource(CachingArchiveClassFileProvider.java:38)
	at com.android.builder.dexing.r8.ClassFileProviderFactory$OrderedClassFileResourceProvider.getProgramResource(ClassFileProviderFactory.java:78)
	at com.android.tools.r8.utils.J.a(:2)
	at com.android.tools.r8.utils.G.a(:8)
	at com.android.tools.r8.utils.G.get(:1)
	at com.android.tools.r8.utils.H.a(:28)
	at com.android.tools.r8.graph.D0.definitionFor(:19)
	at com.android.tools.r8.graph.c.definitionFor(:16)
	at com.android.tools.r8.graph.f0.a(:18)
	at com.android.tools.r8.ir.optimize.d0.b(:79)
	at com.android.tools.r8.ir.conversion.O.a(:587)
	at com.android.tools.r8.ir.conversion.O.a(:417)
	at com.android.tools.r8.ir.conversion.O.b(:53)
	... 53 more

```

I've tried doing some searches around the internet and I haven't really come across and similar issues (especially not ones related to hilt).

Here is my app build.gradle:
```
apply plugin: 'com.android.application'
apply plugin: 'dagger.hilt.android.plugin'

android {}

dependencies {
    // ...

    implementation "com.google.dagger:hilt-android:2.28-alpha"
    implementation 'androidx.hilt:hilt-lifecycle-viewmodel:1.0.0-alpha02'
    annotationProcessor 'androidx.hilt:hilt-compiler:1.0.0-alpha02'
    annotationProcessor "com.google.dagger:hilt-android-compiler:2.28-alpha"
}
```

and project build.gradle:
```
buildscript {
  // ...
  dependencies {
    // ..
    classpath 'com.google.dagger:hilt-android-gradle-plugin:2.28-alpha'
  }
}

// ...
```

Let me know if there's any other info I should provide.

**Comment by bcorso** [comment]
Hi @Allen-HiCharlie,

I don't see anything related to Hilt in the error.

It might be helpful to work backwards to understand at which state in the migration you started seeing the errors. For example, do you see the errors just after adding the Hilt dependencies and the plugin, or only once you start migrating the sources? Also might help to restart AndroidStudio and invalidate caches just to rule that out.

Cheers,
Brad

**Comment by danysantiago** [comment]
This looks very similar to https://github.com/google/dagger/issues/1952. What version of AGP are you on? If on 4.1.0 can you try the latest?

**Comment by Allen-HiCharlie** [comment]
@bcorso Thanks for the advice, the issue occurs immediately after adding the annotation. Prior to that, my app compiles and runs fine. I've also tried invalidating AS cache, `./gradlew clean`, etc but it had no effect 😢 

@danysantiago I've seen #1952 and when I first read it, I did not know what AGP was but now I know it means Android Gradle Plugin.

Following the instructions [here](https://developer.android.com/studio/releases/gradle-plugin) and [here](https://stackoverflow.com/a/53331510) on how to upgrade to 4.1.0, along with [this](https://stackoverflow.com/a/24460241) stackoverflow question to import my existing AS settings, the issue was resolved.

Thanks for your help!
